src/deriveNativeScriptHash.c

```diff
@@ -0,0 +1,469 @@
+{
+	// if the number of remaining scripts is not bigger than 0, then this request
+	// is invalid in the current context, as Ledger was not expecting another
+	// script to be parsed
+	return ctx->complexScripts[ctx->level].remainingScripts > 0;
+}
+
+static inline bool isComplexScriptFinished()
+{
+	return ctx->level > 0 && ctx->complexScripts[ctx->level].remainingScripts == 0;
+}
+
+static inline void complexScriptFinished()
+{
+	while (isComplexScriptFinished()) {
+		ASSERT(ctx->level > 0);
+		ctx->level--;
+
+		ASSERT(ctx->complexScripts[ctx->level].remainingScripts > 0);
+		ctx->complexScripts[ctx->level].remainingScripts--;
+
+		TRACE_WITH_CTX("complex script finished, ");
+	}
+}
+
+static inline void simpleScriptFinished()
+{
+	ASSERT(ctx->complexScripts[ctx->level].remainingScripts > 0);
+	ctx->complexScripts[ctx->level].remainingScripts--;
+
+	TRACE_WITH_CTX("simple script finished, ");
+
+	if (isComplexScriptFinished()) {
+		complexScriptFinished();
+	}
+}
+
+// UI
+
+const char ui_native_script_header[7][27] = {"Script - key path", "Script - key", "Script - ALL", "Script - ANY", "Script - N of K", "Script - invalid before", "Script - invalid hereafter"};
+
+#define ASSERT_UI_SCRIPT_TYPE_SANITY() ASSERT(ctx->ui_scriptType >= UI_SCRIPT_PUBKEY_PATH && ctx->ui_scriptType <= UI_SCRIPT_INVALID_HEREAFTER)
+#define HEADER ui_native_script_header[ctx->ui_scriptType]
+
+static uint8_t _getScriptLevelForPosition()
+{
+	// For complex scripts we reduce the current level by 1
+	// Because they already have the level increased by 1
+	uint8_t levelOffset = ctx->ui_scriptType == UI_SCRIPT_ALL
+	                   || ctx->ui_scriptType == UI_SCRIPT_ANY
+	                   || ctx->ui_scriptType == UI_SCRIPT_N_OF_K
+	                   ? 1 : 0;
+	ASSERT(levelOffset == 0 || ctx->level > 0);
+	return ctx->level - levelOffset;
+}
+
+static void deriveScriptHash_display_ui_position(uint8_t level, ui_callback_fn_t* callback)
+{
+	ASSERT_UI_SCRIPT_TYPE_SANITY();
+	TRACE();
+
+	const size_t POSITION_PREFIX_LENGTH = 11;
+	const size_t MAX_LEVEL_LENGTH = 12;
+
+	// For level n the longest possible position is: "x.x.x...x"
+	// where the x is 2^32-1 (length 10) and there are up to level xs
+	// + the leading "Position: "
+	char positionDescription[10 + 11 * (MAX_SCRIPT_DEPTH - 1) + 1];
+
+	// snprintf returns 0, https://github.com/LedgerHQ/nanos-secure-sdk/issues/28
+	// so we need to check the number of written characters by `strlen`
+	snprintf(positionDescription, POSITION_PREFIX_LENGTH, "Position: ");
+	int offset = strlen(positionDescription);
+
+	for (size_t i = 1; i <= level; i++) {
+		ASSERT(i < MAX_SCRIPT_DEPTH);
+		uint32_t position = ctx->complexScripts[i].totalScripts - ctx->complexScripts[i].remainingScripts + 1;
+		snprintf(&positionDescription[offset], MAX_LEVEL_LENGTH, "%u.", position);
+		offset += strlen(&positionDescription[offset]);
+	}
+
+	// remove any trailing '.'
+	positionDescription[offset - 1] = '\0';
```

Comment:
Let's assert `offset > 0`.

Reply:
Also, it is not obvious that the last char actually is a dot (i.e. why the for cycle is run at least once) --- it depends on the given parameter `level`, and so this function should assert it is given a meaningful value for it.

Reply:
I added an assertion whether the level is bigger than 0.